File: src/CoreFiles/Solver.fs
namespace Trulla.Core

open TheBlunt
open Trulla.Core.Utils

type RecordDef =
    {
        id: TVar
        fields: Field list
        name: string
    }

type Solution =
    {
        solution: Map<TVar, Typ>
        tree: TExp list
        records: RecordDef list
    }

[<RequireQualifiedAccess>]
module Solver =
    let [<Literal>] RootRecordName = "Root"

    let solveParseResult parserResult =
        let mutable usedRecordNames = []

        match parserResult with
        | PError err -> Error [ { range = Range.create err.idx err.idx; message = err.message } ]
        | POk tokens  ->
            match Ast.buildTree tokens.result with
            | Error errs -> Error errs
            | Ok ast ->
                let problems = Inference.buildProblems ast.tree
                match Inference.solveProblems problems ast.tvarToMemberExp with
                | Error errs -> Error errs
                | Ok solution ->
                    let makeRecord tvar fields =
                        let inferedRecordName =
                            match tvar with
                            | Root -> RootRecordName
                            | TVar tvar ->
                                let lastSegmentOfMemberExp =
                                    let memberExp = ast.tvarToMemberExp |> Map.find (TVar tvar) fields
                                    match memberExp with
                                    | IdentExp ident -> ident
                                    | AccessExp accExp -> accExp.memberName
                                let count = 
                                    usedRecordNames 
                                    |> List.filter (fun x -> x = lastSegmentOfMemberExp)
                                    |> List.length
                                do usedRecordNames <- lastSegmentOfMemberExp :: usedRecordNames
                                match count with
                                | 0 -> lastSegmentOfMemberExp
                                | n -> lastSegmentOfMemberExp + (n.ToString())
                        {
                            id = tvar
                            name = inferedRecordName
                            fields = fields
                        }
                    let records =
                        solution
                        |> List.choose (fun (tvar,t) ->
                            match t with
                            | Field f -> Some (tvar,f)
                            | _ -> None)
                        |> List.groupBy fst
                        |> List.map (fun (tvar, fields) -> 
                            fields
                            |> List.map snd 
                            |> List.sortBy (fun f -> f.name)
                            |> makeRecord tvar
                        )
                        |>
                            function
                            | [] -> [ makeRecord Root [] ]
                            | records -> records
                    {
                        solution = solution |> Map.ofList
                        tree = ast.tree
                        records = records 
                    }
                    |> Ok

    let solve template =
        let parseResult = Parsing.parseTemplate template
        solveParseResult parseResult

File: src/NodeCli/lib/Common/CoreFiles/Solver.js
import { Record } from "../../fable_modules/fable-library.4.3.0/Types.js";
import { TVar, AstModule_buildTree, TExp_$reflection, TVar_$reflection } from "./Ast.js";
import { Inference_buildProblems, Inference_solveProblems, Typ_$reflection, Field_$reflection } from "./Inference.js";
import { class_type, record_type, string_type, list_type } from "../../fable_modules/fable-library.4.3.0/Reflection.js";
import { singleton, isEmpty, choose, sortBy, map as map_1, cons, filter, length, empty } from "../../fable_modules/fable-library.4.3.0/List.js";
import { Map_find } from "./Utils.js";
import { compare, safeHash, equals, comparePrimitives, int32ToString } from "../../fable_modules/fable-library.4.3.0/Util.js";
import { List_groupBy } from "../../fable_modules/fable-library.4.3.0/Seq2.js";
import { ofList } from "../../fable_modules/fable-library.4.3.0/Map.js";
import { FSharpResult$2 } from "../../fable_modules/fable-library.4.3.0/Choice.js";
import { Range$ } from "./TheBlunt.js";
import { TrullaError } from "./Core.js";
import { Parsing_parseTemplate } from "./Parsing.js";

export class RecordDef extends Record {
    constructor(id, fields, name) {
        super();
        this.id = id;
        this.fields = fields;
        this.name = name;
    }
}

export function RecordDef_$reflection() {
    return record_type("Trulla.Core.RecordDef", [], RecordDef, () => [["id", TVar_$reflection()], ["fields", list_type(Field_$reflection())], ["name", string_type]]);
}

export class Solution extends Record {
    constructor(solution, tree, records) {
        super();
        this.solution = solution;
        this.tree = tree;
        this.records = records;
    }
}

export function Solution_$reflection() {
    return record_type("Trulla.Core.Solution", [], Solution, () => [["solution", class_type("Microsoft.FSharp.Collections.FSharpMap`2", [TVar_$reflection(), Typ_$reflection()])], ["tree", list_type(TExp_$reflection())], ["records", list_type(RecordDef_$reflection())]]);
}

export function Solver_solveParseResult(parserResult) {
    let usedRecordNames = empty();
    if (parserResult.tag === 0) {
        const matchValue = AstModule_buildTree(parserResult.fields[0].result);
        if (matchValue.tag === 0) {
            const ast = matchValue.fields[0];
            const matchValue_1 = Inference_solveProblems(Inference_buildProblems(ast.tree), ast.tvarToMemberExp);
            if (matchValue_1.tag === 0) {
                const solution = matchValue_1.fields[0];
                const makeRecord = (tvar, fields) => {
                    let lastSegmentOfMemberExp, memberExp, count;
                    return new RecordDef(tvar, fields, (tvar.tag === 1) ? ((lastSegmentOfMemberExp = ((memberExp = Map_find(new TVar(1, [tvar.fields[0]]), fields, ast.tvarToMemberExp), (memberExp.tag === 0) ? memberExp.fields[0].memberName : memberExp.fields[0])), (count = (length(filter((x) => (x === lastSegmentOfMemberExp), usedRecordNames)) | 0), (usedRecordNames = cons(lastSegmentOfMemberExp, usedRecordNames), (count === 0) ? lastSegmentOfMemberExp : (lastSegmentOfMemberExp + int32ToString(count)))))) : "Root");
                };
                let records_1;
                const _arg = map_1((tupledArg_1) => makeRecord(tupledArg_1[0], sortBy((f_1) => f_1.name, map_1((tuple_1) => tuple_1[1], tupledArg_1[1]), {
                    Compare: comparePrimitives,
                })), List_groupBy((tuple) => tuple[0], choose((tupledArg) => {
                    const t = tupledArg[1];
                    if (t.tag === 2) {
                        return [tupledArg[0], t.fields[0]];
                    }
                    else {
                        return void 0;
                    }
                }, solution), {
                    Equals: equals,
                    GetHashCode: safeHash,
                }));
                records_1 = (isEmpty(_arg) ? singleton(makeRecord(new TVar(0, []), empty())) : _arg);
                return new FSharpResult$2(0, [new Solution(ofList(solution, {
                    Compare: compare,
                }), ast.tree, records_1)]);
            }
            else {
                return new FSharpResult$2(1, [matchValue_1.fields[0]]);
            }
        }
        else {
            return new FSharpResult$2(1, [matchValue.fields[0]]);
        }
    }
    else {
        const err = parserResult.fields[0];
        return new FSharpResult$2(1, [singleton(new TrullaError(new Range$(err.idx, err.idx), err.message))]);
    }
}

export function Solver_solve(template) {
    return Solver_solveParseResult(Parsing_parseTemplate(template));
}

